qemu/target/riscv/unicorn.c

```diff
@@ -76,539 +77,106 @@ static void riscv_release(void *ctx)
-    case UC_RISCV_REG_CYCLEH:
-    case UC_RISCV_REG_TIMEH:
-    case UC_RISCV_REG_INSTRETH:
-    case UC_RISCV_REG_HPMCOUNTER3H:
-    case UC_RISCV_REG_HPMCOUNTER4H:
-    case UC_RISCV_REG_HPMCOUNTER5H:
-    case UC_RISCV_REG_HPMCOUNTER6H:
-    case UC_RISCV_REG_HPMCOUNTER7H:
-    case UC_RISCV_REG_HPMCOUNTER8H:
-    case UC_RISCV_REG_HPMCOUNTER9H:
-    case UC_RISCV_REG_HPMCOUNTER10H:
-    case UC_RISCV_REG_HPMCOUNTER11H:
-    case UC_RISCV_REG_HPMCOUNTER12H:
-    case UC_RISCV_REG_HPMCOUNTER13H:
-    case UC_RISCV_REG_HPMCOUNTER14H:
-    case UC_RISCV_REG_HPMCOUNTER15H:
-    case UC_RISCV_REG_HPMCOUNTER16H:
-    case UC_RISCV_REG_HPMCOUNTER17H:
-    case UC_RISCV_REG_HPMCOUNTER18H:
-    case UC_RISCV_REG_HPMCOUNTER19H:
-    case UC_RISCV_REG_HPMCOUNTER20H:
-    case UC_RISCV_REG_HPMCOUNTER21H:
-    case UC_RISCV_REG_HPMCOUNTER22H:
-    case UC_RISCV_REG_HPMCOUNTER23H:
-    case UC_RISCV_REG_HPMCOUNTER24H:
-    case UC_RISCV_REG_HPMCOUNTER25H:
-    case UC_RISCV_REG_HPMCOUNTER26H:
-    case UC_RISCV_REG_HPMCOUNTER27H:
-    case UC_RISCV_REG_HPMCOUNTER28H:
-    case UC_RISCV_REG_HPMCOUNTER29H:
-    case UC_RISCV_REG_HPMCOUNTER30H:
-    case UC_RISCV_REG_HPMCOUNTER31H:
-    case UC_RISCV_REG_MCYCLE:
-    case UC_RISCV_REG_MINSTRET:
-    case UC_RISCV_REG_MCYCLEH:
-    case UC_RISCV_REG_MINSTRETH:
-    case UC_RISCV_REG_MVENDORID:
-    case UC_RISCV_REG_MARCHID:
-    case UC_RISCV_REG_MIMPID:
-    case UC_RISCV_REG_MHARTID:
-    case UC_RISCV_REG_MSTATUS:
-    case UC_RISCV_REG_MISA:
-    case UC_RISCV_REG_MEDELEG:
-    case UC_RISCV_REG_MIDELEG:
-    case UC_RISCV_REG_MIE:
-    case UC_RISCV_REG_MTVEC:
-    case UC_RISCV_REG_MCOUNTEREN:
-    case UC_RISCV_REG_MSTATUSH:
-    case UC_RISCV_REG_MUCOUNTEREN:
-    case UC_RISCV_REG_MSCOUNTEREN:
-    case UC_RISCV_REG_MHCOUNTEREN:
-    case UC_RISCV_REG_MSCRATCH:
-    case UC_RISCV_REG_MEPC:
-    case UC_RISCV_REG_MCAUSE:
-    case UC_RISCV_REG_MTVAL:
-    case UC_RISCV_REG_MIP:
-    case UC_RISCV_REG_MBADADDR:
-    case UC_RISCV_REG_SSTATUS:
-    case UC_RISCV_REG_SEDELEG:
-    case UC_RISCV_REG_SIDELEG:
-    case UC_RISCV_REG_SIE:
-    case UC_RISCV_REG_STVEC:
-    case UC_RISCV_REG_SCOUNTEREN:
-    case UC_RISCV_REG_SSCRATCH:
-    case UC_RISCV_REG_SEPC:
-    case UC_RISCV_REG_SCAUSE:
-    case UC_RISCV_REG_STVAL:
-    case UC_RISCV_REG_SIP:
-    case UC_RISCV_REG_SBADADDR:
-    case UC_RISCV_REG_SPTBR:
-    case UC_RISCV_REG_SATP:
-    case UC_RISCV_REG_HSTATUS:
-    case UC_RISCV_REG_HEDELEG:
-    case UC_RISCV_REG_HIDELEG:
-    case UC_RISCV_REG_HIE:
-    case UC_RISCV_REG_HCOUNTEREN:
-    case UC_RISCV_REG_HTVAL:
-    case UC_RISCV_REG_HIP:
-    case UC_RISCV_REG_HTINST:
-    case UC_RISCV_REG_HGATP:
-    case UC_RISCV_REG_HTIMEDELTA:
-    case UC_RISCV_REG_HTIMEDELTAH: {
+    } else if (regid >= UC_RISCV_REG_F0 &&
+               regid <= UC_RISCV_REG_F31) { // "ft0".."ft31"
+        CHECK_REG_TYPE(uint64_t);
+        *(uint64_t *)value = env->fpr[regid - UC_RISCV_REG_F0];
```

Comment:
note bug fix here (slight incompatibility): F registers should read all 64 bits even in 32-bit mode, as they can hold double-precision floating point values (doubles)

Reply:
This is fine!